Frameworks/CoreGraphics/DWriteWrapper.mm

```diff
@@ -832,35 +398,29 @@ static HRESULT __DWriteUpdateUserCreatedFontCollection(CFArrayRef datas, CFArray
     // Still want to update font collection with whatever fonts succeeded, so return ret at end
     HRESULT ret = func(loader.Get(), datas, errors);
 
-    // DWrite won't create a new font collection unless we provide a new key each time
-    // So every time we modify the font collection increment the key
-    static size_t collectionKey = 0;
-    std::lock_guard<std::mutex> guard(_userCreatedFontCollectionMutex);
-    RETURN_IF_FAILED(
-        dwriteFactory->CreateCustomFontCollection(loader.Get(), &(++collectionKey), sizeof(collectionKey), &_userCreatedFontCollection));
-
-    // Update s_userFontPropertiesMap with new values
-    s_userFontPropertiesMap = __CreatePropertiesMapForFontCollection(_userCreatedFontCollection.Get());
+    // Update the user font collection
+    std::lock_guard<std::recursive_mutex> guard(s_userFontCollectionLock);
```

Comment:
The key thing you said is "lock *around* the whole data structure"

This points to a problem where you don't have the ownership quite correct. The data structure should be able to manage itself (hooray object oriented programming) and therefore the responsibility of locking should be encapsulated to the owning class / object. The pattern as is presents maintainability issues as every new callsite needs to remember to lock otherwise the whole lock is useless. Presumably the problem space for the class itself is much smaller and isolated which makes this a simpler maintainability problem.

---
In reply to: [90937664](https://github.com/Microsoft/WinObjC/pull/1504#discussion_r90937664) [](ancestors = 90937664,90936157)